### ISSUE TITLE
As a user I would like to see incoming mails in my inbox

### ISSUE THREAD

**Issue by fbernitt** [pixelated/pixelated-user-agent#445]
Currently incoming mails are not indexed when they are synced to the user agent. Only after the user agent is restarted, the new incoming mails are availalbe.

In issue #439 the MailboxIndexer got migrated to the new data model. But this is not enough as the current leap_mail code does not yield the newMessages method, instead it gets called like a normal method. See [IMAPMailbox.notify_new](https://github.com/pixelated/leap_mail/blob/develop/src/leap/mail/imap/mailbox.py#L397). The [MailboxIndexer.newMessages](https://github.com/pixelated/pixelated-user-agent/blob/master/service/pixelated/adapter/listeners/mailbox_indexer_listener.py#L38) method gets called but the Deferred is never executed.
Notice: If the user agent gets restartet all mails are indexed and thus also the new mails in the INBOX. But they should be indexed while the user agent is running.

Requirements:
Find a solution that allows us to index incoming mails, either by finding a better hook to listen for incoming mails or by using something like crochet to solve the problem

Given I have an existing pixelated user
And I login into the users account
When I sent a mail to his account by SMTP
And wait until new mails have been fetched
Then I should see this mail in the users inbox INBOX

**Comment by fbernitt** [comment]
It seems at the moment we don't have any test for incoming mails aside from the end-to-end tests we run on staging. Perhaps we can write a incoming mail entry directly to soledad in an integration test context and at least verify that our MailboxIndexer works as expected.

**Comment by shyba** [comment]
I tested this before closing #439 and it worked.
Retested again now and worked again, but I agree with you that lost deferreds are strange. O.o
So, I decided to go deeper and test that.
see: https://paste.debian.net/307948/
When a callback fires a deferred, it is executed.
Our method is a inlineCallback and leap_mail fires our newMessages inside a loop on a non twisted function (notify_new) used as callback.
In this pasted example you can see that callbackWorking is a common python function calling printAsyncWorks in a loop ignoring the result, which is a inlineCallback and returns a deferred. Like ours.
Even the case where I thought it would not work, it worked.
To be on a 100% safe side, It sounds nice to add a call to reactor.callLater to ensure that our deferred chain gets executed even if notify changes on leap side, but I don't think this is necessary since you added a good test for that.
Thoughts?

**Comment by kaeff** [comment]
Thanks for the investigations. If I understand you correctly, the implementation & test is basically ok. For QA, I verified the functionality by sending an email from external, and it arrived, so I'm closing this. I'm unsure how/where the `callLater` would make sense, but I it seems it would be a useful defensive measure, so feel free to add it.

**Comment by fbernitt** [comment]
Interesting, could not reproduce the problem with a new account, too. Guess we can be happy with the test and close this issue for now.
